pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -106,6 +120,53 @@ func haveOverlap(a1, a2 []string) bool {
 	return false
 }
 
+func (pl *VolumeRestrictions) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+	return pl.isReadWriteOncePodAccessModeConflict(pod)
+}
+
+// isReadWriteOncePodAccessModeConflict checks if a pod uses a PVC with the ReadWriteOncePod access mode.
+// This access mode restricts volume access to a single pod on a single node. Since only a single pod can
+// use a ReadWriteOncePod PVC, mark any other pods attempting to use this PVC as unschedulable and unresolvable.
+func (pl *VolumeRestrictions) isReadWriteOncePodAccessModeConflict(pod *v1.Pod) *framework.Status {
+	nodeInfos, err := pl.nodeInfoLister.NodeInfos().List()
+	if err != nil {
+		return framework.NewStatus(framework.Error, ErrReasonNodeInfoError)
+	}
+
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+
+		pvc, err := pl.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(volume.PersistentVolumeClaim.ClaimName)
+		if err != nil {
```

Comment:
check if the error is for NotFound.
If that is the case, the status should be UnschedulableAndUnresolvable. Otherwise, the status is Error.

Reply:
When we create a pod and PVC for the first time, we are likely to hit this `NotFoundError` when we lookup the PVC. It looks like the pod then hangs for ~1m until it retries.

We could use an informer or poll the lister until the PVC is available, however this introduces some extra complexity.

Alternatively, we could query the PVC lister with the existing PVCs on the node, but that's likely less performant because there's more cache lookups and then we would have to enumerate all the pods on the node and check if they're in the right namespace, etc. From what I can tell that removes the need for the `PVCRefCount`.

Reply:
Both when there is an Error and when you return Unschedulable or UnschedulableAndUnresolvable, there is a backoff. This is exponential, and it starts at 1 second.

Returning and leaving it to a retry is the correct thing to do.

We should not return Error when the PVC is not found, as that affects metrics.

Reply:
I'm not able to reproduce this behavior on my kind cluster. This is what the scheduling of a single pod looks like when we return `Error`:
```
Events:
  Type     Reason                  Age                  From                     Message
  ----     ------                  ----                 ----                     -------
  Warning  FailedScheduling        3m56s                default-scheduler        running PreFilter plugin "VolumeRestrictions": failed to locate PersistentVolumeClaim
  Normal   Scheduled               3m55s                default-scheduler        Successfully assigned default/web-server to kind-control-plane
  Normal   SuccessfulAttachVolume  3m54s                attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-ecbb19b1-d55a-4a58-9b26-bf20b9cf48a5"
  Normal   Pulling                 3m53s                kubelet                  Pulling image "nginx"
  Normal   Pulled                  3m48s                kubelet                  Successfully pulled image "nginx" in 5.112477878s
  Normal   Created                 3m46s                kubelet                  Created container web-server
  Normal   Started                 3m46s                kubelet                  Started container web-server
```

And this is what it looks like when it's marked `UnschedulableAndUnresolvable`:
```
Events:
  Type     Reason                  Age                From                     Message
  ----     ------                  ----               ----                     -------
  Warning  FailedScheduling        111s               default-scheduler        0/1 nodes are available: 1 error while getting PersistentVolumeClaim.
  Normal   Scheduled               30s                default-scheduler        Successfully assigned default/web-server to kind-control-plane
  Normal   SuccessfulAttachVolume  30s                attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-597b3026-e66c-4a2d-adac-5e70f1ee86bf"
  Normal   Pulling                 22s                kubelet                  Pulling image "nginx"
  Normal   Pulled                  17s                kubelet                  Successfully pulled image "nginx" in 5.289617494s
  Normal   Created                 15s                kubelet                  Created container web-server
  Normal   Started                 15s                kubelet                  Started container web-server
```

There is a 1 second difference in the case of `Error`, and an 81s difference for `UnschedulableAndUnresolvable`. Is this expected? Currently checking logs for anything that might explain what I'm seeing.

Reply:
That is because on [PreFilter error we don't record the failing plugin](https://github.com/kubernetes/kubernetes/blob/8955463c178636c82ccd94394507ec1c8b67fe2c/pkg/scheduler/core/generic_scheduler.go#L228), and so on any new event it seems that pod is being moved back to the active queue [because of this](https://github.com/kubernetes/kubernetes/blob/53592b322211075fa1f1e98217ab6274e9fc3c64/pkg/scheduler/internal/queue/scheduling_queue.go#L554). 

However, we set the failing plugin on [unschedulable status](https://github.com/kubernetes/kubernetes/blob/8955463c178636c82ccd94394507ec1c8b67fe2c/pkg/scheduler/core/generic_scheduler.go#L234), but sense VolumeBinding plugin is not registering any events, update/create PVC is not triggering moving the pod out of the unschedulable queue. There is a PR for adding that https://github.com/kubernetes/kubernetes/pull/100003

/cc @Huang-Wei 

Wei, I thought plugins that don't register any events is considered wildcard, isn't that the case?

Reply:
It looks like I can then register events on the `VolumeRestrictions` plugin to respond on PVC add or update and this should work, is this right?

Reply:
There is an open PR for that ( #100003), but yes do that in your test with kind to verify that you will get faster response.

Reply:
It looks like this is adding it to the `VolumeBinding` plugin which is different than `VolumeRestrictions`, maybe I'm misunderstanding

Reply:
> Wei, I thought plugins that don't register any events is considered wildcard, isn't that the case?

@ahg-g Plugins that don't register any events will register all default plugins:

https://github.com/kubernetes/kubernetes/blob/3a07d96d2560b491ba6072de4b53d2c97f9ca906/pkg/scheduler/framework/runtime/framework.go#L390-L392

Below is the default plugin list in 1.22 (comparing to 1.21, Service was removed):

https://github.com/kubernetes/kubernetes/blob/3a07d96d2560b491ba6072de4b53d2c97f9ca906/pkg/scheduler/framework/runtime/framework.go#L62-L69

Reply:
> It looks like this is adding it to the VolumeBinding plugin which is different than VolumeRestrictions, maybe I'm misunderstanding

ah sorry. Then yes, you need to add the PVC events to this plugin.

Reply:
Thanks Wei, then this is make sense, we just need to add those events to the current set we have in `VolumeRestrictions`, I was mistakenly looking at `VolumeBinding` plugin

Reply:
I'm a little confused. It seems like `VolumeRestrictions` should have `PersistentVolumeClaim` events registered by default since it currently register any events. [In this case](https://github.com/kubernetes/kubernetes/blob/53592b322211075fa1f1e98217ab6274e9fc3c64/pkg/scheduler/internal/queue/scheduling_queue.go#L554), I would think we eventually find a matching event (the default `PersistentVolumeClaim` one) and proceed. Testing this now.

Reply:
It's working, thank you everyone! (Marking as `UnschedulableAndUnresolvable`)
```
Events:
  Type     Reason                  Age                From                     Message
  ----     ------                  ----               ----                     -------
  Warning  FailedScheduling        31s                default-scheduler        0/1 nodes are available: 1 error while getting PersistentVolumeClaim.
  Normal   Scheduled               30s                default-scheduler        Successfully assigned default/web-server to kind-control-plane
  Normal   SuccessfulAttachVolume  30s                attachdetach-controller  AttachVolume.Attach succeeded for volume "pvc-92c36085-0a26-42f0-b6bd-403abde82cb3"
  Normal   Pulling                 22s                kubelet                  Pulling image "nginx"
  Normal   Pulled                  17s                kubelet                  Successfully pulled image "nginx" in 5.181057731s
  Normal   Created                 15s                kubelet                  Created container web-server
  Normal   Started                 15s                kubelet                  Started container web-server
```

Reply:
Great, can you resolve this comment? way too long and is making it a bit hard to look at the code :)

Reply:
Yes definitely. It's funny, I've always glossed over that button. When I resolve the conversation, are reviewers notified of this?